File: php/test/LiveTest.php
<?php

// Live end-to-end + surface test (PHP): a real headless Chrome session driven
// through the pure-Aether engine via ext-ffi. chromedriver + a content server
// (php's built-in server) are spawned here; FFI calls are synchronous, so the
// content server runs in its own process. markTestSkipped when chromedriver is
// absent, so the suite is green on a box without a browser.

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use SeleniumCore\By;
use SeleniumCore\WebDriver;
use SeleniumCore\NoSuchElementException;

final class LiveTest extends TestCase
{
    private static function which(string $cmd): ?string
    {
        foreach (\explode(':', \getenv('PATH') ?: '') as $dir) {
            $p = $dir . '/' . $cmd;
            if (\is_file($p) && \is_executable($p)) {
                return $p;
            }
        }
        return null;
    }

    private static function freePort(): int
    {
        $s = \stream_socket_server('tcp://127.0.0.1:0', $errno, $errstr);
        $name = \stream_socket_get_name($s, false);
        \fclose($s);
        return (int) \substr($name, \strrpos($name, ':') + 1);
    }

    private static function waitUp(int $port, int $timeoutMs): bool
    {
        $deadline = \microtime(true) + $timeoutMs / 1000.0;
        while (\microtime(true) < $deadline) {
            $c = @\fsockopen('127.0.0.1', $port, $e, $s, 0.5);
            if ($c) {
                \fclose($c);
                return true;
            }
            \usleep(100000);
        }
        return false;
    }

    public function testLiveChromeSurface(): void
    {
        $driverBin = self::which('chromedriver');
        if ($driverBin === null) {
            $this->markTestSkipped('chromedriver not on PATH');
        }

        // Content server: php built-in server over a router that serves two pages.
        $router = \sys_get_temp_dir() . '/sel_php_router_' . \getmypid() . '.php';
        \file_put_contents($router, <<<'PHP'
<?php
$two = '<!doctype html><title>Page Two</title><h1 id="hdr">Two</h1>';
$one = '<!doctype html><title>Page One</title><h1 id="hdr">One</h1>'
     . '<a id="go" href="/two">to two</a>'
     . '<button id="btn" onclick="document.getElementById(\'hdr\').textContent=\'clicked\'">b</button>';
header('Content-Type: text/html; charset=utf-8');
echo str_starts_with($_SERVER['REQUEST_URI'], '/two') ? $two : $one;
PHP);
        $webPort = self::freePort();
        $web = \proc_open(
            [\PHP_BINARY, '-S', "127.0.0.1:$webPort", $router],
            [['pipe', 'r'], ['file', '/dev/null', 'w'], ['file', '/dev/null', 'w']],
            $webPipes
        );
        $base = "http://127.0.0.1:$webPort";

        $cdPort = self::freePort();
        $cd = \proc_open(
            [$driverBin, "--port=$cdPort"],
            [['pipe', 'r'], ['file', '/dev/null', 'w'], ['file', '/dev/null', 'w']],
            $cdPipes
        );

        try {
            if (!self::waitUp($cdPort, 10000) || !self::waitUp($webPort, 5000)) {
                $this->markTestSkipped('chromedriver/server did not come up');
            }

            $d = WebDriver::headlessChrome("http://127.0.0.1:$cdPort");
            try {
                $this->assertGreaterThan(0, \strlen($d->sessionId()), 'session started');

                $d->get("$base/one");
                $this->assertSame('Page One', $d->title(), 'title');
                $this->assertSame('One', $d->findElement(By::ID, 'hdr')->text(), 'hdr text');
                $this->assertSame('a', \strtolower($d->findElement(By::CSS, '#go')->tagName()), 'tag name');

                // navigation
                $d->findElement(By::ID, 'go')->click();
                $this->assertSame('Page Two', $d->title(), 'after click');
                $d->back();
                $this->assertSame('Page One', $d->title(), 'after back');
                $d->forward();
                $this->assertSame('Page Two', $d->title(), 'after forward');
                $d->back();

                // cookies
                $d->deleteAllCookies();
                $d->addCookie(['name' => 'flavor', 'value' => 'mint']);
                $this->assertSame('mint', $d->getCookie('flavor')['value'], 'cookie value');
                $d->deleteCookie('flavor');

                // windows
                $this->assertGreaterThanOrEqual(1, \count($d->windowHandles()), 'window handles');
                $d->setWindowRect(['width' => 900, 'height' => 650]);
                $this->assertSame(900, (int) $d->getWindowRect()['width'], 'window width');

                // script
                $this->assertSame(42, (int) $d->executeScript('return 6*7;'), 'script scalar');
                $this->assertSame(42, (int) $d->executeScript('return arguments[0]+arguments[1];', [40, 2]), 'script args');

                // W3C actions
                $r = $d->findElement(By::ID, 'btn')->rect();
                $cx = (int) ($r['x'] + $r['width'] / 2);
                $cy = (int) ($r['y'] + $r['height'] / 2);
                $d->performActions([[
                    'type' => 'pointer', 'id' => 'mouse',
                    'parameters' => ['pointerType' => 'mouse'],
                    'actions' => [
                        ['type' => 'pointerMove', 'duration' => 0, 'x' => $cx, 'y' => $cy],
                        ['type' => 'pointerDown', 'button' => 0],
                        ['type' => 'pointerUp', 'button' => 0],
                    ],
                ]]);
                $this->assertSame('clicked', $d->findElement(By::ID, 'hdr')->text(), 'actions click fired');
                $d->clearActions();

                // screenshot
                $png = \base64_decode($d->screenshotBase64());
                $this->assertTrue(\strlen($png) > 8 && \substr($png, 1, 3) === 'PNG', 'screenshot is PNG');

                // negative path
                $nse = false;
                try {
                    $d->findElement(By::ID, 'does-not-exist');
                } catch (NoSuchElementException $e) {
                    $nse = true;
                }
                $this->assertTrue($nse, 'no such element error');
            } finally {
                $d->quit();
            }
        } finally {
            if (\is_resource($cd)) {
                \proc_terminate($cd);
            }
            if (\is_resource($web)) {
                \proc_terminate($web);
            }
            @\unlink($router);
        }
    }
}

File: kotlin/src/test/kotlin/org/openqa/selenium/kotlin/LiveTest.kt
package org.openqa.selenium.kotlin

import com.sun.net.httpserver.HttpServer
import org.openqa.selenium.NoSuchElementException
import org.openqa.selenium.RemoteWebDriver
import org.openqa.selenium.WebDriverException
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.util.Base64

// FFI + live surface test for the Kotlin binding. No JUnit — a plain main() with
// assertions, run by kotlin/.tests.ae with `kotlin -classpath ...`. Exercises the
// Kotlin sugar (which is pure Java-interop over the FFM binding → the shared
// engine). Live test needs chromedriver; skips otherwise.

private var failures = 0

private fun check(cond: Boolean, label: String) {
    if (cond) println("  ok: $label") else { println("FAIL: $label"); failures++ }
}

private const val PAGE_ONE =
    "<!doctype html><title>Page One</title><h1 id=\"hdr\">One</h1>" +
        "<a id=\"go\" href=\"/two\">to two</a>" +
        "<button id=\"btn\" onclick=\"document.getElementById('hdr').textContent='clicked'\">b</button>"
private const val PAGE_TWO = "<!doctype html><title>Page Two</title><h1 id=\"hdr\">Two</h1>"

fun main() {
    // ---- FFI (no browser) ----
    check(RemoteWebDriver.route("get") == "POST /session/:sessionId/url", "route get")
    check(RemoteWebDriver.errorCode("no such element") == 17, "errorCode no such element")
    check(RemoteWebDriver.locator("id", "main").contains("*[id="), "locator id rewrite")
    run {
        var threw = false
        try {
            RemoteWebDriver.chrome("http://127.0.0.1:1", null)
        } catch (e: WebDriverException) {
            threw = e.code() == -1
        }
        check(threw, "transport failure -> code -1")
    }

    // ---- live surface ----
    val driverBin = which("chromedriver")
    if (driverBin == null) {
        println("  (live) SKIPPED: chromedriver not on PATH")
    } else {
        liveSurface(driverBin)
    }

    if (failures == 0) {
        println("PASS: Kotlin tests green")
    } else {
        println("FAILED: $failures Kotlin test(s)")
        kotlin.system.exitProcess(1)
    }
}

private fun liveSurface(driverBin: String) {
    val web = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    web.createContext("/") { ex ->
        val body = (if (ex.requestURI.path.startsWith("/two")) PAGE_TWO else PAGE_ONE)
            .toByteArray(StandardCharsets.UTF_8)
        ex.responseHeaders.add("Content-Type", "text/html; charset=utf-8")
        ex.sendResponseHeaders(200, body.size.toLong())
        ex.responseBody.use { it.write(body) }
    }
    web.start()
    val base = "http://127.0.0.1:${web.address.port}"

    val cdPort = freePort()
    val cd = ProcessBuilder(driverBin, "--port=$cdPort")
        .redirectOutput(ProcessBuilder.Redirect.DISCARD)
        .redirectError(ProcessBuilder.Redirect.DISCARD)
        .start()
    try {
        if (!waitUp(cdPort, 10_000)) {
            println("  (live) SKIPPED: chromedriver did not come up")
            return
        }
        headlessChrome("http://127.0.0.1:$cdPort") { d ->
            check(d.sessionId().isNotEmpty(), "session started")

            d.get("$base/one")
            check(d.getTitle() == "Page One", "title")
            check(d.find(By.id("hdr")).getText() == "One", "hdr text")
            check(d.find(By.cssSelector("#go")).getTagName().lowercase() == "a", "tag name")

            // navigation
            d.find(By.id("go")).click()
            check(d.getTitle() == "Page Two", "after click")
            d.back(); check(d.getTitle() == "Page One", "after back")
            d.forward(); check(d.getTitle() == "Page Two", "after forward")
            d.back()

            // cookies
            d.deleteAllCookies()
            d.addCookie(mapOf("name" to "flavor", "value" to "mint"))
            @Suppress("UNCHECKED_CAST")
            val cookie = d.getCookie("flavor") as Map<String, Any?>
            check(cookie["value"] == "mint", "cookie value")
            d.deleteCookie("flavor")

            // windows
            check(d.windowHandles().isNotEmpty(), "window handles")
            d.setWindowRect(mapOf("width" to 900, "height" to 650))
            @Suppress("UNCHECKED_CAST")
            val rect = d.getWindowRect() as Map<String, Any?>
            check((rect["width"] as Number).toInt() == 900, "window width")

            // script shapes
            check((d.script("return 6*7;") as Number).toInt() == 42, "script scalar")
            check((d.script("return arguments[0]+arguments[1];", 40, 2) as Number).toInt() == 42, "script args")

            // W3C actions: pointer click on the button
            @Suppress("UNCHECKED_CAST")
            val br = d.find(By.id("btn")).rect() as Map<String, Any?>
            val cx = ((br["x"] as Number).toDouble() + (br["width"] as Number).toDouble() / 2).toInt()
            val cy = ((br["y"] as Number).toDouble() + (br["height"] as Number).toDouble() / 2).toInt()
            d.performActions(
                listOf(
                    mapOf(
                        "type" to "pointer", "id" to "mouse",
                        "parameters" to mapOf("pointerType" to "mouse"),
                        "actions" to listOf(
                            mapOf("type" to "pointerMove", "duration" to 0, "x" to cx, "y" to cy),
                            mapOf("type" to "pointerDown", "button" to 0),
                            mapOf("type" to "pointerUp", "button" to 0),
                        ),
                    ),
                ),
            )
            check(d.find(By.id("hdr")).getText() == "clicked", "actions click fired")
            d.clearActions()

            // screenshot -> PNG
            val png = Base64.getDecoder().decode(d.screenshotBase64())
            check(png.size > 8 && png[1] == 'P'.code.toByte() && png[2] == 'N'.code.toByte(), "screenshot is PNG")

            // negative path
            var nse = false
            try {
                d.find(By.id("does-not-exist"))
            } catch (e: NoSuchElementException) {
                nse = true
            }
            check(nse, "no such element error")
        }
    } finally {
        cd.destroy()
        web.stop(0)
    }
}

private fun which(cmd: String): String? {
    val path = System.getenv("PATH") ?: return null
    for (dir in path.split(":")) {
        val f = java.io.File(dir, cmd)
        if (f.canExecute() && !f.isDirectory) return f.absolutePath
    }
    return null
}

private fun freePort(): Int = ServerSocket(0, 0, java.net.InetAddress.getByName("127.0.0.1")).use { it.localPort }

private fun waitUp(port: Int, timeoutMs: Long): Boolean {
    val deadline = System.currentTimeMillis() + timeoutMs
    while (System.currentTimeMillis() < deadline) {
        try {
            Socket("127.0.0.1", port).close()
            return true
        } catch (_: Exception) {
            Thread.sleep(100)
        }
    }
    return false
}
